src/native/managed/cdacreader/src/SOSDacImpl.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+
+namespace Microsoft.Diagnostics.DataContractReader;
+
+[GeneratedComInterface]
+[Guid("4eca42d8-7e7b-4c8a-a116-7bfbf6929267")]
+internal partial interface ISOSDacInterface9
+{
+    void GetBreakingChangeVersion(out int version);
+}
+
+/// <summary>
+/// Implementation of ISOSDacInterface* interfaces intended to be passed out to consumers
+/// interacting with the DAC via those COM interfaces.
+/// </summary>
+[GeneratedComClass]
+internal sealed partial class SOSDacImpl : ISOSDacInterface9
+{
+    private readonly Target _target;
+
+    public SOSDacImpl(Target target)
+    {
+        _target = target;
+    }
+
+    public void GetBreakingChangeVersion(out int version)
+    {
+        // TODO: Return non-hard-coded version
```

Comment:
This API represents versions of set of algorithmic and data contracts that SOS and CLRMD happens to depend on. It should eventually obsoleted, and SOS and CLRMD switched to use the versions of properly defined contracts.

(It is ok to keep it for now if it helps.)

Reply:
Are we trying to implement the actual C++ ISOSDacInterface9 interface here (`HRESULT GetBreakingChangeVersion(int* pVersion)`)? Or is this part of the new C# cDAC interface being defined?

Reply:
> Are we trying to implement the actual C++ ISOSDacInterface9 interface here ? Or is this part of the new C# cDAC interface being defined?

We're trying to implement the actual ISOSDacInterface9 here.  This is going to be consumed by the existing DAC in cases where it will delegate to the cdacreader.

>  Or is this part of the new C# cDAC interface being defined?

It is not a goal right now to define a public interface to the cDAC

Reply:
Then shouldn't the C# method be ` public int GetBreakingChangeVersion(out int version)`? Or is there some magic (code generation/COM interop) that will expose the C++ function `HRESULT GetBreakingChangeVersion(int* pVersion)`?

Reply:
>  that will expose the C++ function HRESULT GetBreakingChangeVersion(int* pVersion)?

COM interop has elided the `HRESULT` by default since at least .NET Framework 2.0. We can have a matching signature as you've suggested, but the `out` isn't correct either since it is unmanaged memory, so it should be `int*`. I would prefer we either match the signature precisely using blittable types or we write it in canonical .NET form. In this case that means using the last "out" param as the return and letting the marshalling system throw an `Exception` for a failing `HRESULT` or convert an `Exception` to an `HRESULT` in the other direction.

The one caveat for this is when `S_FALSE` is used. In cases where an API can return a non-zero success `HRESULT` we need to use the `PreserveSig` mechanism and check the `int` manually.